src/build/opBuildAPIClient.ts

```diff
@@ -0,0 +1,87 @@
+import { EventStream } from '../common/EventStream';
+import { extensionConfig } from '../shared';
+import { APICallFailed, APICallStarted, CredentialExpired } from '../common/loggingEvents';
+
+const OP_BUILD_USER_TOKEN_HEADER_NAME = 'X-OP-BuildUserToken';
+
+export class OPBuildAPIClient {
+    constructor(private environmentController: EnvironmentController) { }
+
+    public async getOriginalRepositoryUrl(gitRepoUrl: string, opBuildUserToken: string, eventStream: EventStream): Promise<string> {
+        const requestUrl = `${this.APIBaseUrl}/v2/Repositories/OriginalRepositoryUrl`
+            + `?${querystring.stringify({
+                gitRepoUrl
+            })}`;
+
+        return this.sendRequest('GetOriginalRepositoryUrl', requestUrl, 'GET', opBuildUserToken, eventStream)
+            .then((response: any) => {
+                if (this.isResultEmpty(response)) {
+                    throw new Error(`Cannot the get original repository URL for current repository(The repository is not provisioned in Docs).`);
+                }
+                return response.toString();
+            }, (code: any) => {
+                throw new Error(`Cannot the get original repository URL for current repository(${code})`);
+            });
+    }
+
+    private get APIBaseUrl() {
+        return extensionConfig.OPBuildAPIEndPoint[this.environmentController.env];
+    }
+
+    private async sendRequest(
+        name: string,
+        url: string,
+        method: string,
+        token: string,
+        eventStream: EventStream,
+        reqObj: any = {},
+        acceptedStatusCode: number[] = [200],
+        headers: any = {},
+    ): Promise<any> {
+        method = method || 'GET';
+        headers[OP_BUILD_USER_TOKEN_HEADER_NAME] = token;
+
+        eventStream.post(new APICallStarted(name, url));
+        const promise = new Promise((resolve, reject) => {
+            request({
+                url,
+                method,
+                headers,
+                json: true,
+                body: reqObj,
+            }, (error, response, respObj) => {
+                if (error) {
+                    eventStream.post(new APICallFailed(name, url, error.message));
+                    reject();
+                }
+
+                if (!acceptedStatusCode.includes(response.statusCode)) {
+                    if (response.statusCode === 401) {
+                        eventStream.post(new CredentialExpired());
+                    }
+
+                    eventStream.post(new APICallFailed(name, url, `${respObj.error ? (`${respObj.error}(${response.statusCode}): `) : 'error: '}`
+                        + `${respObj.message ? respObj.message : JSON.stringify(respObj)}`));
+                    reject(response.statusCode);
+                }
+                resolve(respObj);
+            });
+        });
+        return promise;
+    }
+
+    private isResultEmpty(result: any): boolean {
+        if (Array.isArray(result)) {

```

Comment:
should the result type be known already or this will be shared with later usage?

Reply:
hope it can be shared in the future.

---
In reply to: [365078703](https://github.com/docascode/vscode-docs-build/pull/13#discussion_r365078703) [](ancestors = 365078703)